feat(linter): overhaul unicorn/no-useless-spread

DonIsaac opened the pull request:
I got tired of seeing useless spreads on ternaries and `arr.reduce()` within my company's internal codebase so I overhauled this rule.

## Changes
- add fixer for object spreads
  ```js
  const before = { a, ...{ b, c }, d }
  const after = { a,  b, c, d } // fixer does not dedupe spaces before `b`
  ```
- recursively check for useless clones on complex expressions. This rule now catches and auto-fixes the following cases:
   ```js
  // ternaries when both branches create a new array or object
   const obj = { ...(foo ? { a: 1 } : { b: 2 }) }
  // recursive, so this can support complex cases
  const arr = [ ...(foo ? a.map(fn) : bar ? Array.from(iter) : await Promise.all(bar)) ]
  // reduce functions where the initial accumulator creates a new object or array
   const obj = { ...(arr.reduce(fn, {}) }
  ```

mysteryven commented on crates/oxc_linter/src/rules/unicorn/no_useless_spread/mod.rs:
```diff
@@ -326,153 +338,141 @@ fn check_useless_iterable_to_array<'a>(
-            },
-        );
+    let hint = target.const_eval();
+    let hint_matches_expr = if is_array { hint.is_array() } else { hint.is_object() };
```
What's the means of `hint_matches_expr` here? 🤔

DonIsaac replied:
true if
1. we're checking for clones on arrays and `ValueHint` suggests the value is an array
2. vice versa for objects

DonIsaac committed:
feat(linter): overhaul unicorn/no-useless-spread (#4791)

I got tired of seeing useless spreads on ternaries and `arr.reduce()` within my company's internal codebase so I overhauled this rule.

## Changes
- add fixer for object spreads
  ```js
  const before = { a, ...{ b, c }, d }
  const after = { a,  b, c, d } // fixer does not dedupe spaces before `b`
  ```
- recursively check for useless clones on complex expressions. This rule now catches and auto-fixes the following cases:
   ```js
  // ternaries when both branches create a new array or object
   const obj = { ...(foo ? { a: 1 } : { b: 2 }) }
  // recursive, so this can support complex cases
  const arr = [ ...(foo ? a.map(fn) : bar ? Array.from(iter) : await Promise.all(bar)) ]
  // reduce functions where the initial accumulator creates a new object or array
   const obj = { ...(arr.reduce(fn, {}) }
  ```
